### ISSUE TITLE
When hosting as Windows Service, Singletons are disposed AFTER the Service has reported to be stopped

### ISSUE THREAD

**Issue by lucas-natraj** [dotnet/runtime#63284]
### Description

(duplicate of https://github.com/dotnet/aspnetcore/issues/39264)

Pretty much what the title describes. 
It appears that the Service does wait for `IHostedService` implementations to finish before reporting the status as stopped, but NOT for `IDisposable` (or `IASyncDisposable`) singletons.

This has an awkward side-effect when Restarting the WindowsService as the previous process singleton instances may still be around when new singletons are instantiated in the newly launched process.

### Reproduction Steps

The repository at https://github.com/lucas-natraj/dueling-services represents a minimal setup to reproduce the issue.

1. build the solution -> resulting in the executable binary `Lastly.exe`
1. register the exe as a windows service using `sc.exe create Lastly binPath= {path-to-lastly.exe}`
1. install [seq](https://datalust.co/download) for view aggregated logs. or monitor the `log***.txt` files in the executable directory.
1. start the service (via TaskManager)
1. hit the webapi endpoint by visiting `http://localhost:5000/ping`.
1. `Restart` the window service (via TaskManager)
1. view logs

The aggregated logs in `Seq` show the service reporting as stopped before the `Ping` singleton is disposed.
Additionally it can be seen that a new service instance is launched before the singleton from the previous instance has been disposed.
Note that the `hosted` service is appropriately 'cleaned up' before the service reports as stopped.

<img width="1084" alt="seq-logs" src="https://user-images.githubusercontent.com/7196683/147871871-d8ac8fdc-e363-41fe-a456-bfe2e0205f82.png">

### Expected behavior

I would have expected that the service would wait until all singletons are disposed (within some timeout-limit) before reporting the Service as Stopped.

### Actual behavior

Singletons are disposed AFTER the service has reported itself as Stopped.
In the case of a `TaskManager -> Service Restart` tends to result in Windows launching another instance of the process when the previous one has not yet terminated, and various Singletons yet to be disposed.

### Configuration

dotnet --info

```
.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.1
  Commit:  3a25a7f1cc

.NET SDKs installed:
  5.0.401 [C:\Program Files\dotnet\sdk]
  5.0.402 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.2.21505.57 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21480.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21501.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by lucas-natraj** [comment]
[This line](https://github.com/dotnet/runtime/blob/7574e03e71d35827bfc08ece73492bd3df21c61d/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L62) results in marking the service as `Stopped` as soon as the `ApplicationLifetime.ApplicationStopped` cancellation token is triggered.

The ServiceProvider is only disposed AFTER ApplicationStopped is triggered.

One potential solution is if we rely solely on [WindowsServiceLifetime.Dispose](https://github.com/dotnet/runtime/blob/7574e03e71d35827bfc08ece73492bd3df21c61d/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L122) to mark the service as Stopped.

**Comment by eerhardt** [comment]
@Tratcher - any thoughts on removing 

https://github.com/dotnet/runtime/blob/7574e03e71d35827bfc08ece73492bd3df21c61d/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L60-L63

and relying on `Dispose` to unblock `OnStop`?

**Comment by Tratcher** [comment]
Relying on WindowsServiceLifetime.Dispose doesn't guarantee all services have been disposed, only that one. That might be close enough to the desired behavior since I think services are disposed in reverse resolution order.

A singleton service has no cross-process lifetime guarantees. E.g., you'd also have to handle this scenario if your process crashed and Dispose was not run.

Relying on Dispose could also result in more timeouts here if there are any errors during shutdown.
https://github.com/dotnet/runtime/blob/7574e03e71d35827bfc08ece73492bd3df21c61d/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L106

**Comment by lucas-natraj** [comment]
thanks for the quick response @Tratcher.

> Relying on WindowsServiceLifetime.Dispose doesn't guarantee all services have been disposed, only that one. That might be close enough to the desired behavior since I think services are disposed in reverse resolution order.

the execution flow is a little involved, so I'm not too sure, but I believe under normal circumstances the dispose is called through ServiceBase.
https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.ServiceProcess/ServiceBase.cs#L761-L766

but you're right, it does get us much closer to the desired result.

> A singleton service has no cross-process lifetime guarantees. E.g., you'd also have to handle this scenario if your process crashed and Dispose was not run.

if the process crashed, then all resources will be 'cleaned up' anyway (albeit in a more brutal fashion), the Service will be detected as toast, and the system will restart the Windows Service / process automatically (if configured as such). 

> Relying on Dispose could also result in more timeouts here if there are any errors during shutdown.
> 
> https://github.com/dotnet/runtime/blob/7574e03e71d35827bfc08ece73492bd3df21c61d/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L106

yes, although would that be a problem? 
in the failure case, the service would just wait that `ShutdownTimeout` (default 5 sec, i believe) before marking as Stopped.
and a longer timeout could be specified if a consumer would like to extend the wait for a particularly lengthy tear-down scenario.

note that it'll be quite unlikely to correctly handle all the exceptional cases, so I'm really just trying to more closely align the Windows Service Running / Stopped boundaries as close as possible to the lifespan of the application (and its sub-components). 
as of right now it is ALWAYS wrong. 😆 

any other suggestions on what we could do?

**Comment by tthiery** [comment]
A question: is it a valid expectation to correlate these two states (dispose and service stopped). After all, the process is still alive (as it is with windows services lifetime) and dispose is not an alternative for an orderly shutdown (see file close vs dispose).

Personally I would also correlate them, but the .NET pattern landscape imho insist otherwise.

**Comment by lucas-natraj** [comment]
@tthiery that's a reasonable question, and under normal circumstances it's likely easier to keep these concerns separate.

the [WindowsServiceLifetime](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs) class tries to bridge the .net Hosting lifecycle with that of the windows service infrastructure in [ServiceBase](https://github.com/microsoft/referencesource/blob/master/System.ServiceProcess/ServiceBase.cs).

as such, there are various hooks that can be used on both sides of this bridge to achieve an appropriate alignment.
on the .net Hosting side there are the various IHostApplicationLifetime tokens and on the Windows Service side there are base class `On...` overrides (like `OnStart, OnStop, OnShutDown` and `Dispose`).

note that the `Dispose` on the WindowsServiceLifetime is called BOTH during the disposal of the DI container, **as well as part of the normal flow of the ServiceBase `Run` method.** 

https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.ServiceProcess/ServiceBase.cs#L761-L766

> .. After all, the process is still alive (as it is with windows services lifetime) 

not sure I understand. the issue is that there is a mismatch between the process and the service status. the service status (via the WindowsServiceLifetime) currently reports that it is Stopped when it is still very much alive.

i'm all for finding an alternate 'hook' for when we declare that the `Service` has, for all intents and purposes, stopped. 
my initial suggestion was just an initial stab at something relatively easy.

**Comment by eerhardt** [comment]
It feels like there are some layering issues to be addressed here on how we should think of the Extensions.Hosting system w.r.t. the Windows ServiceBase layer. Does Extensions.Hosting sit on top of ServiceBase? On the side? Who is the true "owner" of the process?

Related, see the discussion in https://github.com/dotnet/runtime/issues/62579#issuecomment-996342998. There we want the `ServiceBase.Run` method to finish completely before the process has exited. When looking at a .NET Framework "Windows Service" templated app, it has:

```C#
        static void Main()
        {
            ServiceBase[] ServicesToRun;
            ServicesToRun = new ServiceBase[]
            {
                new Service1()
            };
            ServiceBase.Run(ServicesToRun);
        }
```

Which means the `ServiceBase.Run` method will definitely complete before the process can cleanly exit.

With all this in mind, it appears to me that we think of Extensions.Hosting as sitting "on top of" ServiceBase and is the ultimate "owner" of the process. After all, it is what is being "Run" in Main in the template.

Given that, I think it is correct to say that the "Host" object doesn't get disposed until after the Service has been stopped. (Note that Singleton services get disposed when the "Host" object gets disposed.) The "Host" object is the owner/controller of the process. It doesn't make sense to dispose of it before the Windows Service has been stopped, because it is the thing that is running the Windows Service. So in my mind, the ordering looks like this:

1. Host is created and Run
2. ServiceBase.Run is called as part of running the Host
3. ....
4. Something triggers the application to be stopped.
5. The Windows Service is "stopped"
6. The Host is disposed
7. The process exits

@Tratcher @ericstj - thoughts on this analysis?

**Comment by lucas-natraj** [comment]
Just a quick reminder that, right now, `IHostedService` instances are ’terminated’ BEFORE the Service is Stopped.

It’s only the DI container that is, for some reason, done after.

**Comment by eerhardt** [comment]
That makes sense. Note the separation between `IHostedService` and `IHost`. The `IHostedService` instances are what runs the logic of the service. The `IHost` is what orchestrates the whole process.

> It’s only the DI container that is, for some reason, done after.

That is because the DI container is disposed as part of disposing the `IHost`.

**Comment by Tratcher** [comment]
@eerhardt sounds about right. Starting, Stopping, and Disposing are all distinct steps. If there's a resource you need torn down on stop rather than dispose then do that in the IHostedService.StopAsync call.

**Comment by lucas-natraj** [comment]
it seems that we have a 'Host' that comprises of HostedServices (background / long-running tasks, ...) and associated Service objects (registered in the DI container that assist in doing whatever).
i'm not following why we would draw the 'disposed' line AFTER HostedService but BEFORE the other (singleton or otherwise) services?

A very slightly modified sequence would look something like:

```
1. host is created and run
|-- 2. ServiceBase.Run is called -----------------------------> windows service reports as 'Started'.
|-- 3. hosted services launched
|-- 4. singletons and other helper objects instantiated upon demand
|-- ... (application running) ...
(something triggers the application to be stopped.)
|-- 5. hosted services are shutdown
|-- 6. singletons / other instances disposed
|-- 7. ServiceBase is stopped --------------------------------> windows service reports as 'Stopped'.
|-- 8. any final host cleanup / disposal (if any) is done
9. process exits
```

we would still maintain the 'layering' with the Host encapsulating the lifespan of the windows service, but in this case, the time window between the service reporting as stopped (7.) and the processes exiting (9.) would be likely insignificant.

most users would expect that the stopping of a service would align very closely to the tear-down and stopping of the associated process.

note that I'm not particularly concerned (at least in this comment) about where stop / disposal logic goes, but rather in trying to more closely align the Started / Stopped status of the Windows Service with that of the Host.

**Comment by eerhardt** [comment]
A thing to understand is that the singleton services' lifetime is tied to the DI container (i.e. `IServiceProvider`). And the DI container's lifetime is tied to the Host object. Disposing the Host object is what disposes the DI container, which is what disposes the singleton services. You can't dispose of them earlier, while the Host is still valid, because someone could be trying to use one of those services during the "Stop" process (ex. `IHostApplicationLifetime.ApplicationStopped`). So you can't split (6) and (8) up in your sequence above.

**Comment by lucas-natraj** [comment]
yeah, i have dug through the [Host](https://github.com/dotnet/runtime/blob/00952e17c49a5666a58e871758a4fadcdd2b0a33/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs) code, so I do appreciate your point.

that list was more a 'desired' sequence of events that a 'user' would expect.
the difficultly, as I see it, is whether we can get close to that expected sequence without breaking the universe. 😄 

as of now, it may be several seconds between when the service shows up as stopped and things actually going away.

not to mention that weird side-effect where client code always needs to protect itself from multiple duplicate process instances, instead of relying on the WindowsService infrastructure to take care of it.

**Comment by tthiery** [comment]
I am actually a not invested user. I just bumbled over this ticket. I see the world like that: dispose activities i care about and the ones I do not care. I care about everything explicitly which holds resources: ports, files, timers, ... i need to plan them. In Java, .NET, C, C++, ...

So when I write a service where I care about the resources, i clean them up explicitly. If I do not do this, there are consequences like processes hanging around after the expected lifetime. So for me, this is not a bug considering the hierarchy process, host, DI, service/hostedservice.

If we see this on a higher level than managed 
code programming, then the question is: why there is not one DI Container per service or hostedservice instance. But I think this ship has sailed and is also not desirable. Maybe something like a static (not per request) DI scope per service is something interesting to explore. This could be then disposed in sync with the service runtime.

**Comment by lucas-natraj** [comment]
> So when I write a service where I care about the resources, i clean them up explicitly. If I do not do this, there are consequences like processes hanging around after the expected lifetime. So for me, this is not a bug considering the hierarchy process, host, DI, service/hostedservice.

i understand the sentiment here, but relying on IDisposable IS a reasonable way of cleaning things up. It is well documented, and the Host code honors disposable objects.

It seems a little awkward to force consumers, in this scenario, to MOVE everything to a HostedService if they want disposal logic to run before the WindowsService shows up as Stopped.

Our application is somewhat large, and has seen some significant (albeit) slow migrations at modernization. When our services were direct inheritors of ServiceBase, there was never an issue of duplicate processes, or a significant delay until the process truly went away.

It's only when we wanted to take advantage of all the modern .net hosting capabilities and started using the WindowsServiceLifetime have we stumbled into this unexpected behavior.

Sure, we can easily work around this, but it seems that other people might also follow the same path we have, and it'd be nice if the Windows Service integration worked closer to how it has historically functioned.

**Comment by maryamariyan** [comment]
closing as it appears the question has been answered as to behavior is the way it is.

**Comment by lucas-natraj** [comment]
uh, i'm not sure why this issue is closed because the behavior 'as-is' isn't particularly ideal. 
we might have a way to explain why it is currently failing, but surely there something that we can do to improve this.

i'd expect more issues to arise in future around the current timing discrepancy. (see potentially https://github.com/dotnet/runtime/issues/62579)

as a side note, we have been running the potential fix described above for a few weeks now in production with significantly better cleanup / disposal success than earlier.
